File: grpc-trading-platform/user-service/src/main/java/org/hle/user/util/EntityMessageMapper.java
package org.hle.user.util;

import org.hle.user.Holding;
import org.hle.user.StockTradeRequest;
import org.hle.user.StockTradeResponse;
import org.hle.user.UserInformation;
import org.hle.user.entity.PortfolioItem;
import org.hle.user.entity.User;

import java.util.List;

public class EntityMessageMapper {

    public static UserInformation toUserInformation(User user, List<PortfolioItem> items){
        var holding = items.stream()
                .map(i -> Holding.newBuilder()
                        .setTicker(i.getTicker())
                        .setQuantity(i.getQuantity())
                        .build())
                .toList();

        return UserInformation.newBuilder()
                .setUserId(user.getId())
                .setName(user.getName())
                .setBalance(user.getBalance())
                .addAllHoldings(holding)
                .build();
    }

    public static PortfolioItem toPortfolioItem(StockTradeRequest request) {
        var item = new PortfolioItem();
        item.setUserId(request.getUserId());
        item.setTicker(request.getTicker());
        item.setQuantity(request.getQuantity());

        return item;
    }

    public static StockTradeResponse toStockTradeResponse(StockTradeRequest request, int balance) {
        return StockTradeResponse.newBuilder()
                .setUserId(request.getUserId())
                .setPrice(request.getPrice())
                .setTicker(request.getTicker())
                .setQuantity(request.getQuantity())
                .setAction(request.getAction())
                .setTotalPrice(request.getPrice() * request.getQuantity())
                .setBalance(balance)
                .build();
    }
}

File: grpc-trading-official/user-service/src/main/kotlin/org/hle/userservice/util/EntityMessageMapper.kt
package org.hle.userservice.util

import org.hle.userservice.entity.User
import org.hle.userservice.entity.PortfolioItem
import org.hle.user.Holding
import org.hle.user.StockTradeRequest
import org.hle.user.StockTradeResponse
import org.hle.user.UserInformation

object EntityMessageMapper {
    fun toUserInformation(user: User, items: kotlin.collections.MutableList<PortfolioItem?>): UserInformation {
        val holding = items.stream()
            .map<Holding?> { i: PortfolioItem? ->
                Holding.newBuilder()
                    .setTicker(i?.ticker)
                    .setQuantity(i?.quantity ?: 0)
                    .build()
            }
            .toList()

        return UserInformation.newBuilder()
            .setUserId(user.id ?: 0)
            .setName(user.name)
            .setBalance(user.balance ?: 0)
            .addAllHoldings(holding)
            .build()
    }

    fun toPortfolioItem(request: StockTradeRequest): PortfolioItem {
        val item: PortfolioItem = PortfolioItem()
        item.userId = request.getUserId()
        item.ticker = request.getTicker()
        item.quantity = request.getQuantity()

        return item
    }

    fun toStockTradeResponse(request: StockTradeRequest, balance: kotlin.Int): StockTradeResponse {
        return StockTradeResponse.newBuilder()
            .setUserId(request.getUserId())
            .setPrice(request.getPrice())
            .setTicker(request.getTicker())
            .setQuantity(request.getQuantity())
            .setAction(request.getAction())
            .setTotalPrice(request.getPrice() * request.getQuantity())
            .setBalance(balance)
            .build()
    }
}
